keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,362 @@
+- Service Discovery: Support for both A and AAAA records for kube-DNS and coreDNS
+- Functionality tested with the Bridge CNI plugin and Host-Local IPAM plugin as references
+- Maintain backwards-compatible support for IPv4-only and IPv6-only clusters
+
+### Non-Goals
+
+- Cross-family connectivity: IPv4-to-IPv6 and IPv6-to-IPv4 connectivity is considered outside of the scope of this proposal, except in the case of ingress access (depending on ingress controller used).
+- CNI network plugins other than the Bridge and Host-Local IPAM plugins should support Kubernetes dual stack, but the development and testing of dual stack support for these other plugins is considered outside of the scope of this proposal.
+- Code changes should support any number IPv4 addresses and any number of IPv6 addresses per pod (up to some reasonable limit). However, testing will be done only up to the following limits:
+  - Pod addresses: 1 IPv4 address and 2 IPv6 addresses per pod maximum
+  - Service addresses: 1 IPv4 and 1 IPv6 service addresses per service maximum
+- For simplicity, for the first release of dual-stack support, the [Kubernetes Endpoints API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpoints-v1-core) will not be modified. Each IPv4/IPv6 address that is assigned to a backend pod will be treated and displayed as a separate endpoint for the corresponding service. Service load balancing (e.g. using kube-proxy) will be performed on a per-address basis, rather than being done on a per-dual-stack-endpoint basis.
+
+## Proposal
+
+In order to support dual-stack in Kubernetes clusters, Kubernetes needs to have awareness of and support dual-stack addresses for pods, nodes and services. To make this happen, the following changes will be required:
+
+- Kubernetes needs to be made aware of multiple IPs per pod (tested with up to one IPv4 and up to two IPv6 addresses).
+- Link Local Addresses (LLAs) on a pod will remain implicit (Kubernetes will not display nor track these addresses).
+- Kubernetes needs to be configurable for up to two service CIDRs.
+- Backend pods for a service can be dual stack. For the first release of dual-stack support, each IPv4/IPv6 address of a backend pod will be treated as a separate Kubernetes endpoint.
+- Kube-proxy needs to support IPv4 and IPv6 services in parallel (e.g. drive iptables and ip6tables in parallel).
+- Health/liveness/readiness probes should work for dual-stack pods. For the first release of dual-stack support, no additional configuration "knobs" will be added for probe definitions, and a probe is deemed successful if either an IPv4 or IPv6 response is received
+- Kubectl commands and output displays will need to be modified for dual-stack.
+- Kubeadm support will need to be added to enable spin-up of dual-stack clusters.
```

Comment:
I think it's fine to select an orchestrator for the reference implementation. My question is whether the KubeAdm changes should be part of this effort or a companion effort. My vote is to keep the scope of an already large effort as small as possible.

Reply:
Can we develop dual stack, along with CI, without Kubeadm changes?

Reply:
I'm not saying that orchestrator changes are not needed. I'm suggesting breaking up the work into separate efforts ("a companion effort"). Are we going to scope out kubeadm changes in detail as part of this KEP, then?

Reply:
If it's a required part of implementing and testing the dual stack feature, then I'm proposing that we need to mention that here. Specifically, I want to avoid saying that the kubeadm changes are out-of-scope, out of concern that the decoupled kubeadm dual-stack changes might get a lower priority relative to the rest of the dual-stack work. This would be in the spirit of a KEP intended to be a "feature, and effort tracking document" ([from KEP summary](https://github.com/kubernetes/community/blob/master/keps/0001-kubernetes-enhancement-proposal-process.md#summary)). For kubeadm implementation details, we can always file an issue against kubeadm,